lib/spell-check-manager.coffee

```diff
@@ -0,0 +1,306 @@
+    console.log "spell-check: addPluginChecker:", checker
+    @addSpellChecker checker
+  addSpellChecker: (checker) ->
+    console.log "spell-check: addSpellChecker:", checker
+    @checkers.push checker
+
+  removeSpellChecker: (spellChecker) ->
+    @checkers = @checkers.filter (plugin) -> plugin isnt spellChecker
+
+  check: (args, text) ->
+    # Make sure our deferred initialization is done.
+    @init()
+
+    # We need a couple packages.
+    multirange = require 'multi-integer-range'
+
+    # For every registered spellchecker, we need to find out the ranges in the
+    # text that the checker confirms are correct or indicates is a misspelling.
+    # We keep these as separate lists since the different checkers may indicate
+    # the same range for either and we need to be able to remove confirmed words
+    # from the misspelled ones.
+    correct = new multirange.MultiRange([])
+    incorrects = []
+
+    for checker in @checkers
+      # We only care if this plugin contributes to checking spelling.
+      if not checker.isEnabled() or not checker.providesSpelling(args)
+        continue
+
+      # Get the results which includes positive (correct) and negative (incorrect)
+      # ranges.
+      results = checker.check(args, text)
+
+      if results.correct
+        for range in results.correct
+          correct.appendRange(range.start, range.end)
+
+      if results.incorrect
+        newIncorrect = new multirange.MultiRange([])
+        incorrects.push(newIncorrect)
+
+        for range in results.incorrect
+          newIncorrect.appendRange(range.start, range.end)
+
+    # If we don't have any incorrect spellings, then there is nothing to worry
+    # about, so just return and stop processing.
+    misspellings = []
+
+    if incorrects.length is 0
+      return {id: args.id, misspellings}
+
+    # Build up an intersection of all the incorrect ranges. We only treat a word
+    # as being incorrect if *every* checker that provides negative values treats
+    # it as incorrect. We know there are at least one item in this list, so pull
+    # that out. If that is the only one, we don't have to do any additional work,
+    # otherwise we compare every other one against it, removing any elements
+    # that aren't an intersection which (hopefully) will produce a smaller list
+    # with each iteration.
+    intersection = null
+    index = 1
+
+    for incorrect in incorrects
+      if intersection is null
+        intersection = incorrect
+      else
+        intersection.intersect(incorrects[index])
+
+    # If we have no intersection, then nothing to report as a problem.
+    if intersection.length is 0
+      return {id: args.id, misspellings}
+
+    # Remove all of the confirmed correct words from the resulting incorrect
+    # list. This allows us to have correct-only providers as opposed to only
+    # incorrect providers.
+    if correct.ranges.length > 0
+      intersection.subtract(correct)
```

Comment:
Nope, I read it wrong mainly because we have two types of ranges which are based on two entirely different sets. :) The `spellchecker` doesn't use `Range([startRow, startCol], [endRow, endCol])`. Instead, it [returns an array of character ranges](https://github.com/atom/spell-check/blob/master/lib/spell-check-handler.coffee#L7) `[[startIndex, endIndex], ...]`. There was logic to convert those text indexes into row/col ranges [below it](https://github.com/atom/spell-check/blob/master/lib/spell-check-handler.coffee#L9-L30).

I went with the text indexes approach to match `spellchecker` which why I went with `multi-integer-range`. Once I have a single range of incorrect words, then I [convert them to row/col ranges](https://github.com/dmoonfire/spell-check/blob/master/lib/spell-check-manager.coffee#L89-L129).

I expect a single checker so not return overlapping words but the code handles it even if it does. I do expect two checkers to include overlapping ranges which is why I do the intersection stuff.

I'm find with word-based API, it would solve this nicely, but I want to set up a quick interop test. I want to see if there is a noticeable overhead on 10k 1-word calls verses a single 10k word call. And to see if caching would make sense in this case (most languages use the same 30-40 words repeatedly).